### PR TITLE
[FIX] web,mail: Call on_attach_callback when appending content

### PR DESCRIPTION
**Pull request by Arcasias**
Task [2346540](https://www.odoo.com/web#id=2346540&model=project.task&view_type=form&cids=1&menu_id=)

### CODE REVIEW

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 1]
```diff
@@ -1067,15 +1067,18 @@ var FormRenderer = BasicRenderer.extend({
...
-        }
         this.$el.toggleClass('o_form_editable', this.mode === 'edit');
         this.$el.toggleClass('o_form_readonly', this.mode === 'readonly');
 
+        const $content = this.has_sheet ?
+            $(this._wrapFormContents(...$newContent)) :
+            $newContent;
+        dom.append(this.$el, $content, {
+            in_DOM: this._isInDom,
+            callbacks: [{ widget: this }],
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js:1080]: this is incorrect: if `this._isInDom` is true, `on_attach_callback` will be called on the renderer, even though it's already in the DOM. Could you confirm?

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js [hunk 2]
```diff
@@ -167,6 +167,22 @@ FormRenderer.include({
...
+    _wrapFormContents(...els) {
+        const chatterElIndex = els.findIndex(
+            el => el.classList.contains("o_FormRenderer_chatterContainer")
+        );
+        if (chatterElIndex < 0) {
+            return this._super(...els);
+        } else {
+            const [chatterEl] = els.splice(chatterElIndex, 1);
+            const wrapped = this._super(...els);
+            wrapped.splice(chatterElIndex, 0, chatterEl);
```
Comment (aab-odoo) [inline addons/mail/static/src/widgets/form_renderer/form_renderer.js:182]: this is worth a bit of explanation in the docstring

**Review by aab-odoo — COMMENTED** [review]
I think there is a problem with the current solution (see comments). Also, there is no test?

File: addons/web/static/src/js/views/form/form_renderer.js [hunk 3]
```diff
@@ -1019,17 +1019,10 @@ var FormRenderer = BasicRenderer.extend({
...
-        // render the form and evaluate the modifiers
-        var defs = [];
-        this.defs = defs;
-        this.inactiveNotebooks = [];
-        var $form = this._renderNode(this.arch).addClass(this.className);
-        delete this.defs;
-
-        return Promise.all(defs).then(function () {
+        const $form = await this.__renderView();
+        return Promise.resolve().then(function () {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/form/form_renderer.js:1025]:
```suggestion
        return this.__renderView().then(function ($form) {
```

File: addons/web/static/tests/views/form_tests.js [hunk 4]
```diff
@@ -9666,9 +9669,88 @@ QUnit.module('Views', {
...
+            "form.on_attach_callback", // Form attached
+            "widget.on_attach_callback", // Initial widget attached
+            "component.mounted", // Initial component mounted
+        ]);
+
+        await testUtils.form.clickEdit(form);
+
+        assert.verifySteps([
+            "widget.on_attach_callback", // New widget attached
+            "component.mounted", // Component remounted
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form_tests.js:9741]: I don't think this is correct. Component is mounted twice??

File: addons/web/static/src/js/views/form/form_renderer.js (hunk 3 above)
Comment (Arcasias) [reply addons/web/static/src/js/views/form/form_renderer.js:1025]: `$form` is needed in the `guardedCatch` block. I could make it a dynamic var but the line count would be the same.

Comment (aab-odoo) [reply addons/web/static/src/js/views/form/form_renderer.js:1025]: alright, read too fast

**Review by aab-odoo — APPROVED** [review]
For me it's ok. But maybe @seb-odoo or @alexkuhn would like to take a look before we r+.

**Comment by seb-odoo** [comment]
Just asking to be sure, `_renderView` override in enterprise shouldn't also be `__renderView` now?

**Comment by seb-odoo** [comment]
Runbot is red it seems btw.

**Comment by Arcasias** [comment]
> Just asking to be sure, `_renderView` override in enterprise shouldn't also be `__renderView` now?

@seb-odoo `__renderView` was needed because of the asynchronous work performed in the `form_renderer` override which put a delay between the appending of the new content and the `on_attach_callback` calls. We could indeed override `__renderView` in enterprise for consistency but it is not needed, thus I do not think it will should done in stable.

**Comment by aab-odoo** [comment]
Where are we with this @Arcasias @seb-odoo ?

**Comment by seb-odoo** [comment]
The mail part is ok for me, it's just a move of override, as long as it still works...
I let you decide for the rest.

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo retry

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo retry

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by seb-odoo** [comment]
> > Just asking to be sure, `_renderView` override in enterprise shouldn't also be `__renderView` now?
> 
> @seb-odoo `__renderView` was needed because of the asynchronous work performed in the `form_renderer` override which put a delay between the appending of the new content and the `on_attach_callback` calls. We could indeed override `__renderView` in enterprise for consistency but it is not needed, thus I do not think it will should done in stable.

Following discussion today with @aab-odoo I think you should do it in enterprise too.
It's not only about async work done in the override, but about also having the result of the override available before calling the "then" that does the `_updateView` if I'm not mistaken.

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**